crates/wasi-common/src/ctx.rs

```diff
@@ -268,6 +268,19 @@ impl WasiCtxBuilder {
         self
     }
 
+    pub fn preopened_handle<P: AsRef<Path>>(
+        &mut self,
+        handle: Box<dyn Handle>,
+        guest_path: P,
+    ) -> &mut Self {
+        let preopen = PendingPreopen::new(move || Ok(handle));
+        self.preopens
+            .as_mut()
+            .unwrap()
+            .push((guest_path.as_ref().to_owned(), preopen));
+        self
+    }
```

Comment:
Would it be possible to reimplement `preopened_dir` in terms of this function?

Reply:
That would require making the `OsDir::try_from(dir)` part happen directly in `preopened_dir`, rather than in the callback - which in turn requires either modifying the interface to return Result, or panicking (boo!), or making `self.preopens` be a collection of `Result<PendingPreopen, ...>` (which would be... weird).

Of those, I'd rather modify the interface to return result. Reporting errors earlier is generally better.

Reply:
(That said, this feels like a distinct effort from just adding this one method)

Reply:
Makes sense; thanks for looking into it!